lib/mpl_toolkits/mplot3d/axes3d.py

```diff
@@ -1957,6 +1957,108 @@ def plot(self, xs, ys, *args, zdir='z', **kwargs):
 
     plot3D = plot
 
+    def fill_between(self, x1, y1, z1, x2, y2, z2, *,
+                     where=None, mode='auto', **kwargs):
+        """
+        Fill the area between two 3D curves.
+
+        The curves are defined by the points (*x1*, *y1*, *z1*) and
+        (*x2*, *y2*, *z2*). This creates one or multiple quadrangle
+        polygons that are filled. All points must be the same length N, or a
+        single value to be used for all points.
+
+        Parameters
+        ----------
+        x1, y1, z1 : float or 1D array-like
+            x, y, and z  coordinates of vertices for 1st line.
+
+        x2, y2, z2 : float or 1D array-like
+            x, y, and z coordinates of vertices for 2nd line.
+
+        where : array of bool (length N), optional
+            Define *where* to exclude some regions from being filled. The
+            filled regions are defined by the coordinates ``pts[where]``,
+            for all x, y, and z pts. More precisely, fill between ``pts[i]``
+            and ``pts[i+1]`` if ``where[i] and where[i+1]``. Note that this
+            definition implies that an isolated *True* value between two
+            *False* values in *where* will not result in filling. Both sides of
+            the *True* position remain unfilled due to the adjacent *False*
+            values.
+
+        mode : {'auto', 'quad', 'polygon'}, default: 'auto'
+            The fill mode.
+            If 'auto' and all the coordinates for the x, y, or z input pairs
+            are the same, then the points lie on a plane and 'polygon' is used.
+            Otherwise, 'auto' defaults to 'quad'.
+            If 'quad', then a separate quadrilateral polygon is created for
+            each pair of subsequent points in the 1st and 2nd lines. This is
+            more flexible, but may be slower and result in rendering artifacts.
+            If 'polygon', then the 1st and 2nd lines are connected to form a
+            single polygon, which may be faster and render more cleanly for
+            simple shapes (eg, for filling between two lines that lie on a
+            single plane).
+
+        **kwargs
+            All other keyword arguments are passed on to `.Poly3DCollection`.
+
+        Returns
+        -------
+        `.Poly3DCollection`
+            A `.Poly3DCollection` containing the plotted polygons.
+
+        """
+        _api.check_in_list(['auto', 'quad', 'polygon'], mode=mode)
+
+        had_data = self.has_data()
+        x1, y1, z1, x2, y2, z2 = cbook._broadcast_with_masks(x1, y1, z1, x2, y2, z2)
+        if mode == 'auto':
+            if np.all(x1 == x2) or np.all(y1 == y2) or np.all(z1 == z2):
```

Comment:
> The more I think about it, the more I like the optimization to map auto to polygon when all the points are on an x, y, or z plane. It doesn't cover all the cases of points on an arbitrary plane, but it will cover what I think will be the most common ones, and the user can always choose the mode directly.

How hard would it be to generalize to "if all points lie in a plane"? You could construct a plane from the first 3 points and then calculate a dot product for all following to check distance to that plane and it wouldn't be _too_ bad.

Reply:
This is working, but is a little clunky... thoughts?

'''python
        i_nonnan = np.where(~np.isnan(x1))[0]
        xs = np.concatenate((x1[i_nonnan], x2[i_nonnan]))
        ys = np.concatenate((y1[i_nonnan], y2[i_nonnan]))
        zs = np.concatenate((z1[i_nonnan], z2[i_nonnan]))

        # calculate the normal vector from the first three points
        n = np.cross([xs[1] - xs[0], ys[1] - ys[0], zs[1] - zs[0]],
                     [xs[2] - xs[0], ys[2] - ys[0], zs[2] - zs[0]])
        # calculate the vectors from the first point to all other points
        vs = np.array([xs - xs[0], ys - ys[0], zs - zs[0]])
        # if the dot product of the normal vector and all other vectors is zero,
        # all points are on the same plane
        dots = np.dot(n, vs)
        all_points_on_plane = np.allclose(dots, 0, atol=1e-12)
'''

Reply:
@greglucas I added `art3d._all_points_on_plane()` to check for this, it's a little involved to catch the edge cases - what do you think? I still need to add tests but wanted to get a pulse first.

Reply:
Sorry for sending you down the rabbit hole 😂 . I think it is great! and I really appreciate that you checked the edge cases. Maybe it is a little involved, but seems worth it to me to generalize the statement to any plane. It is only a one-time performance hit when creating the collection initially and it is all numpy-vectorized so it isn't that terrible.